src/components/postgresConnect.ts

```diff
@@ -1,45 +1,231 @@
+} from '../types'
 
-const { Client } = require('pg')
+class PostgresDB {
+  private static _instance: PostgresDB
+  private pool: Pool
 
-// Postgres Database listeners for Triggers/Actions
-export const pgClient = new Client(config.pg_database_connection)
+  constructor() {
+    this.pool = new Pool(config.pg_database_connection)
 
-pgClient.connect()
-console.log('Connecting to Postgres...')
+    // the pool will emit an error on behalf of any idle pools
+    // it contains if a backend error or network partition happens
+    this.pool.on('error', (err) => {
+      console.error('Unexpected error on idle pool', err)
+      process.exit(-1)
+    })
 
-// Load action plugins
-const actionLibrary: ActionLibrary = {}
-const actionSchedule: any[] = []
+    this.startListener()
+  }
 
-export const loadActionPlugins = async () => {
-  // Scan plugins folder and update Database
-  await registerPlugins(pgClient)
-  // Load Action functions into global scope
-  await loadActions(pgClient, actionLibrary)
-  // Schedule future actions and execute overdue ones
-  await loadScheduledActions(pgClient, actionLibrary, actionSchedule)
-}
+  private startListener = async () => {
+    const listener = new Client(config.pg_database_connection)
+    listener.connect()
+    listener.query('LISTEN trigger_notifications')
+    listener.query('LISTEN action_notifications')
+    listener.on('notification', ({ channel, payload }) => {
+      if (!payload) {
+        console.log(`Notification ${channel} received with no payload!`)
+        return
+      }
+      switch (channel) {
+        case 'trigger_notifications':
+          processTrigger(JSON.parse(payload))
+          break
+        case 'action_notifications':
+          executeAction(JSON.parse(payload), actionLibrary)
+          break
+      }
+    })
+  }
+
+  private getValuesPlaceholders = (object: { [key: string]: any }) =>
+    Object.keys(object).map((key, index) => `$${index + 1}`)
+
+  public static get Instance() {
+    return this._instance || (this._instance = new this())
+  }
+
+  public query = async (text: string, payload: any[] = []): Promise<QueryResult> => {
+    const client = await this.pool.connect()
+    try {
+      return await client.query(text, payload)
+    } finally {
+      // Make sure to release the client before any error handling,
+      // just in case the error handling itself throws an error.
+      client.release()
+    }
+  }
+
+  public addActionQueue = async (action: ActionQueuePayload): Promise<boolean> => {
+    try {
+      const result = await this.query(
+        `INSERT into action_queue (${Object.keys(
+          action
+        )}, time_queued) VALUES (${this.getValuesPlaceholders(action)}, CURRENT_TIMESTAMP)`,
+        Object.values(action)
+      )
+      return true
+    } catch (err) {
+      console.log(err.stack)
+      return false
+    }
+  }
+
+  public executeActionQueued = async (payload: ActionQueueExecutePayload): Promise<boolean> => {
```

Comment:
Perhaps this function name is somewhat unclear. How about `executedActionStatusUpdate` ?

Reply:
Cool. I'll update it now.